### PR TITLE
Updating Starbucks again

### PR DESCRIPTION
**Pull request by reuning**
I previously submitted the pull request #4706 to get more Starbucks data but I think it had the side effect of taking way too long to run and so no Starbucks data has been collected since. 

I've tried something a bit different this time. Now if it is still searching after the normal step through it will identify stores that it collected that are furthest away from each other and use the long/lat of those in the next search. My initial test tripled the time in searching NYC, but found more stores. This compares to the previous pull request which more than 25x-ed the length of time.

### CODE REVIEW

**Comment by Cj-Malone** [comment]
This API is annoying me. Why give us `paging` if we can't do anything with it?!?

**Review by Cj-Malone — APPROVED** [review]
So I've just been playing with this. I have no way around the current API call :( (@starbucks help us out?)

But I ran the currently merged one (excluding Canada), and after more than a day I encountered some bug in my network stack and so I aborted. It haden't even got past California, starting from the west! 4,601 POIs.

I ran it "plain" just lat/lon search without any of the extra paging bits, again excluding Canada. 10,301 POIs, quick enough that I didn't time it.

And I ran this one, excluding Canada. 17,452 POIs. It took 6 hours, but I had 20% of the requests cached by this point.

```python
{'atp/brand/Starbucks': 17450,
 'atp/brand_wikidata/Q37158': 17450,
 'atp/category/amenity/cafe': 17450,
 'atp/field/email/missing': 17450,
 'atp/field/image/missing': 17450,
 'atp/field/opening_hours/missing': 17450,
 'atp/field/phone/invalid': 10,
 'atp/field/phone/missing': 321,
 'atp/field/postcode/missing': 1,
 'atp/field/twitter/missing': 17450,
 'atp/nsi/match_failed': 17450,
 'downloader/request_bytes': 11127372,
 'downloader/request_count': 22222,
 'downloader/request_method_count/GET': 22222,
 'downloader/response_bytes': 1478032978,
 'downloader/response_count': 22222,
 'downloader/response_status_count/200': 22222,
 'dupefilter/filtered': 20907,
 'elapsed_time_seconds': 21646.501861,
 'feedexport/success_count/FileFeedStorage': 1,
 'finish_reason': 'finished',
 'finish_time': datetime.datetime(2023, 6, 8, 4, 23, 25, 529980),
 'httpcache/firsthand': 17748,
 'httpcache/hit': 4474,
 'httpcache/miss': 17748,
 'httpcache/store': 17748,
 'item_dropped_count': 803919,
 'item_dropped_reasons_count/DropItem': 803919,
 'item_scraped_count': 17450,
 'log_count/DEBUG': 74491,
 'log_count/INFO': 369,
 'log_count/WARNING': 803919,
 'memusage/max': 272961536,
 'memusage/startup': 142381056,
 'request_depth_max': 4,
 'response_received_count': 22222,
 'robotstxt/request_count': 1,
 'robotstxt/response_count': 1,
 'robotstxt/response_status_count/200': 1,
 'scheduler/dequeued': 22221,
 'scheduler/dequeued/memory': 22221,
 'scheduler/enqueued': 22221,
 'scheduler/enqueued/memory': 22221,
 'start_time': datetime.datetime(2023, 6, 7, 22, 22, 39, 28119)}
```

The PR is clearly better than the merged one, so I think we should merge. However we are getting 0 POIs in the weekly, I expect that's a data centre thing, as it worked locally (GB ip). So I don't expect this one to work in the weekly.

**Comment by reuning** [comment]
For what it is worth, most sources I can find point to there [being around 16,000 stores in the US.](https://starbmag.com/starbucks-locations-in-the-united-states/) The old version definitely was missing things. It is too bad this can't work in the weekly pull, I had hoped it was an issue with it running so slowly that it caused issues and this faster version would make it possible.

**Comment by Cj-Malone** [comment]
And actually, even though I was excluding the CA search points, POIs near the border come in anyway.

`$ cat starbucks-new.geojson | jq '.features[].properties."addr:country"' | sort | uniq -c | sort -rn`
```
  15940 "US"
   1434 "CA"
     44 "MX"
     31 "PR"
      1 "DO"
```

So this is basically 100% of POIs :)

> weekly

We'll see once it's merged, but I think it's one of the ones where we'll need a proxy.

**Comment by Cj-Malone** [comment]
> Out of all the stores in the United States nearly nine thousand stores are corporate owned stores while the rest are licensed stores.
> https://starbmag.com/starbucks-locations-in-the-united-states/

I had wondered what `ownership_type=CO` and `ownership_type=LS` meant.

**Comment by iandees** [comment]
I would really like to find a way to do this without numpy. That's a huge dependency to add.

File: locations/spiders/starbucks.py [hunk 1]
```diff
@@ -75,22 +76,61 @@ def parse(self, response):
 
         paging = responseJson["paging"]
         if paging["returned"] > 0 and paging["limit"] == paging["returned"]:
-            if response.meta["distance"] > 0.10:
+            if response.meta["distance"] > 0.15:
                 nextDistance = response.meta["distance"] / 2
-                nextDistanceCorner = nextDistance * (sqrt(2) / 2)
                 # Create eight new coordinate pairs
```
Comment (iandees) [inline locations/spiders/starbucks.py:81]:
```suggestion
                # Create four new coordinate pairs
```

**Comment by reuning** [comment]
I've removed numpy. It isn't perhaps the most eloquent solution but it seems to work.

**Comment by iandees** [comment]
Thanks!

**Comment by reuning** [comment]
Excited that the weekly pull worked, and returned a total of 17,448 locations. Woo.

**Pull request merged**